src/ir/auth_logic/ast_construction_test.cc

```diff
@@ -26,51 +26,76 @@
 #include <variant>
 #include <vector>
 
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
 #include "src/common/testing/gtest.h"
 #include "src/ir/auth_logic/ast.h"
 #include "src/ir/datalog/program.h"
 
 namespace raksha::ir::auth_logic {
 
+using ::testing::UnorderedElementsAre;
+
 struct QueryTestData {
   std::string input_text;
   std::string query_name;
-  std::string principal_name;
-  std::string fact_name;
-  datalog::Predicate predicate;
+  Principal principal;
+  Fact fact;
 };
 
+std::string ToString(datalog::Predicate predicate) {
+  return absl::StrCat(predicate.name(), "(",
+                      absl::StrJoin(predicate.args(), ", "), ")");
+}
+std::string ToString(BaseFact basefact) {
+  if (basefact.GetValue().index() == 0) {  // Predicate BaseFact
+    return absl::StrCat(
+        ToString(std::get<datalog::Predicate>(basefact.GetValue())));
+  } else if (basefact.GetValue().index() == 1) {  // Attribute BaseFact
+    Attribute attribute = std::get<Attribute>(basefact.GetValue());
+    return absl::StrCat(attribute.principal().name(), " ",
+                        ToString(attribute.predicate()));
+  } else {  // CanActAs BaseFact
+    CanActAs canactas = std::get<CanActAs>(basefact.GetValue());
+    return absl::StrCat(canactas.left_principal().name(), " canActAs ",
+                        canactas.right_principal().name());
+  }
+}
+std::string ToString(Fact fact) {
+  std::string cansay_string = "";
+  for (const auto& cansay : fact.cansay()) {
+    cansay_string =
+        absl::StrCat(cansay.principal().name(), " canSay ", cansay_string);
+  }
```

Comment:
Don't repeatedly `StrCat` in a loop like this, especially not to the front of a string. This is likely to be quadratic in runtime. Use `StrJoin` instead.

Reply:
StrJoin needs a separator which in our case is an empty string. This change also might make code not look as redundant.

Reply:
It looks to me like your separator is `" canSay "`

Reply:
Changed to StrJoin wherever possible.